### PR TITLE
move account delete functionality to new data model

### PR DESCRIPTION
**Pull request by kddejong**
## Proposed changes

- Move Account Deletion into the new data model

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (changes to code, which do not change application behavior)

## Checklist

- [ ] I have filled out this PR template
- [ ] I have read the [CONTRIBUTING](https://github.com/Optum/dce/blob/master/docs/CONTRIBUTING.md) doc
- [ ] I have added automated tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (`README.md`, inline comments, etc.)
- [ ] I have updated the `CHANGELOG.md` under a `## next` release, with a short summary of my changes

### CODE REVIEW

File: pkg/api/error.go [hunk 1]
```diff
@@ -31,10 +36,12 @@ func WriteAPIErrorResponse(w http.ResponseWriter, err error) {
 // WriteAPIResponse writes the response out to the provided ResponseWriter
 func WriteAPIResponse(w http.ResponseWriter, status int, body interface{}) {
 	w.WriteHeader(status)
-	err := json.NewEncoder(w).Encode(body)
-	if err != nil {
-		log.Printf("error encoding and writing message: %+v", body)
-		w.WriteHeader(http.StatusInternalServerError)
-		_ = json.NewEncoder(w).Encode(errors.NewInternalServer("error writing response", err))
+	if body != nil {
```
Comment (kddejong) [inline pkg/api/error.go]: Don't write the body if the interface is nil

File: pkg/data/account.go [hunk 2]
```diff
@@ -87,7 +87,7 @@ func (a *Account) Delete(account *account.Account) error {
 			// Query in Lease Table
 			TableName: aws.String(a.TableName),
 			// Return the updated record
-			ReturnValues: aws.String("ALL_NEW"),
+			ReturnValues: aws.String("NONE"),
```
Comment (kddejong) [inline pkg/data/account.go]: Can't be ALL_NEW.  Wasn't using this before so found the issue now.

File: pkg/account/model.go [hunk 3]
```diff
@@ -90,6 +92,32 @@ func (a *Account) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// UnmarshalDynamoDBAttributeValue handles custom unmarshaling of an ARN
+func (a *Account) UnmarshalDynamoDBAttributeValue(av *dynamodb.AttributeValue) error {
+
+	type Alias Account
```
Comment (kddejong) [inline pkg/account/model.go]: If anyone is wondering why this exists like this.  Its to prevent infinite loops with line 101.

**Comment by kddejong** [comment]
Completes the rewrite for #165 for the accounts controller

File: cmd/lambda/accounts/delete.go [hunk 4]
```diff
@@ -1,113 +1,28 @@
...
-	// json.NewEncoder(w).Encode(response.CreateAPIGatewayResponse(http.StatusNoContent, ""))
-	response.WriteAPIResponse(w, http.StatusNoContent, "")
-}
-
-// sendSNS sends notification to SNS that the delete has occurred.
-func sendSNS(account *db.Account) {
-	serializedAccount := response.AccountResponse(*account)
-	serializedMessage, err := common.PrepareSNSMessageJSON(serializedAccount)
 
+	acct, err := Services.Config.AccountService().Get(accountID)
```
Comment (eschwartz) [inline cmd/lambda/accounts/delete.go]: Might be worth calling out in the `Get()` method what kind of errors it returns
(I just had to dig in a bit to figure out that we're actually generating different HTTP error types at the data layer)

File: pkg/account/model.go (hunk 3 above)
Comment (eschwartz) [reply pkg/account/model.go]: Is this `UnmarshalDynamoDBAttributeValue()` method used anywhere? My IDE couldn't find any usages...

File: pkg/accountmanager/service.go [hunk 5]
```diff
@@ -86,6 +86,42 @@ func (s *Service) UpsertPrincipalAccess(account *account.Account) error {
 	return nil
 }
 
+// DeletePrincipalAccess removes all the principal roles and policies
+func (s *Service) DeletePrincipalAccess(account *account.Account) error {
```
Comment (eschwartz) [inline pkg/accountmanager/service.go]: Is this new behavior?

File: pkg/account/service.go [hunk 6]
```diff
@@ -215,6 +218,26 @@ func (a *Service) Delete(data *Account) error {
 		return err
 	}
 
+	err = a.managerSvc.DeletePrincipalAccess(data)
+	if err != nil {
+		return err
+	}
+
+	err = a.dataSvc.Delete(data)
```
Comment (eschwartz) [inline pkg/account/service.go]: Looks like this is a duplicate call, see line 216

File: pkg/account/model.go (hunk 3 above)
Comment (kddejong) [reply pkg/account/model.go]: yes, its an interface used by dynamodbattribute.Unmarshal

File: pkg/account/service.go (hunk 6 above)
Comment (kddejong) [reply pkg/account/service.go]: Yikes good catch.

File: pkg/accountmanager/service.go (hunk 5 above)
Comment (kddejong) [reply pkg/accountmanager/service.go]: Moved code from role and policy manager.

**Pull request merged**